fix(router): stop emitting to event observable on destroy

crisbeto opened the pull request:
No longer emits to `Router.events` after the router has been destroyed. Also  returns a resolved promise to the navigation methods.

Fixes #40502.

atscott reviewed (approved):
I think that with this change, the `navigate` and `navigateByUrl` will return a promise that will never resolve or reject. 
What do you think of also adding a private `disposed` flag that `scheduleNavigation` short-circuits and returns a rejected promise if `dispose` has been called? 
Using the returned `Promise`, especially the rejection of the promise, is very uncommon and the overlap with people navigating after `dispose` is likely 0 so I'm okay leaving this extra logic out as well.

crisbeto commented:
That's a good point I hadn't really considered. I think the problem with returning a blank promise is that it's basically a silent error. In other areas of the framework (e.g. `ChangeDetectorRef`) we throw an error if you try to interact with something that has been destroyed, but adding an error now could be considered a breaking change.

atscott commented:
@crisbeto - so as it turns out, completing these observables would have to be considered a breaking change as well. Specifically, completing the `events` observable can result in tests throwing `EmptyError` if there is a location in the code that has a `ThrowIfEmptySubscriber` on the events (such as the `first()` operator). If the condition isn't met in the lifecycle of a test, this will result in a new `EmptyError` and the test will fail.
This is likely what we would want anyways, but we'll have to at least go through and clean these all up in g3 first and add the breaking change note.

I do wonder if the `Promise.reject` would need to be considered breaking though. If it weren't, we could at least submit that change as a quick fix now, and then add the `complete` for v12.

crisbeto commented:
Regarding the `events` completion: another route we go down is to just stop emitting to it and leave it up to consumers to unsubscribe themselves. Then we can actually complete it in a major version.

As for the `Promise.reject`, I don't think that it's a breaking change, but I was wondering whether it wouldn't make more sense to throw an error when you're interacting with a destroyed router. Throwing the error would be a breaking change IMO.

atscott commented:
> Regarding the `events` completion: another route we can go down is to just stop emitting to it and leave it up to consumers to unsubscribe themselves. Then we can actually complete it in a major version.
> 
> As for the `Promise.reject`, I don't think that it's a breaking change, but I was wondering whether it wouldn't make more sense to throw an error when you're interacting with a destroyed router. Throwing the error would be a breaking change IMO.

Let's go with this way for now in the minor/patch and we can look at completing the observables and throwing an error when interacting with the disposed Router in v12. Immediately returning a rejected promise from `scheduleNavigation` would have the effect of no longer emitting to `events` because those only happen through route transitions.

crisbeto commented:
@atscott I've reworked it so it isn't breaking. One note is that I ended up returning a `Promise.resolve(false)` rather than a `Promise.reject()`, because the rejection could be considered a breaking change since most JS engines throw for uncaught promise rejections.

crisbeto committed:
fix(router): stop emitting to event observable on destroy

No longer emits to `Router.events` after the router has been destroyed. Also
returns a resolved promise to the navigation methods.

Fixes #40502.

atscott commented:
[updated presubmit](https://test.corp.google.com/OCL:355036350:BASE:355486901:1612392349957:1e1e5b64)
